### ISSUE TITLE
better keyboard support in ttyx mode

### ISSUE THREAD

**Issue by unxed** [elfmz/far2l#1126]
Ctrl+1 works as "record macro" instead of "go to bookmark 1 folder" in ttyx mode

Both for RightCtrl and LeftCtrl.

Without ttyx it works just as "1".

Maybe some bug with processing modifiers state from X?

**Comment by unxed** [comment]
Ctrl+2 also works strange in ttyx mode: it makes Ctrl "sticked" for the one next keypress. Other number keys behave strange with Ctrl in ttyx also.

**Comment by unxed** [comment]
btw, can we use XkbGetState in ttyx broker to distinguish left and right ctrls?

**Comment by elfmz** [comment]
well, starting macro seems fixed, however all Ctrl+DIGIT keys  not working properly except Ctrl+1. At least under Gnome terminal, and not sure how to fix that. Check this code: https://www.paste.org/120735 - its a test app that prints what terminal sends to it (to exit from it - press Enter). My gnome term sends '1' on pressing Ctrl+1 while Ctrl+other digits produces insane stuff, like NUL characters, \1b (escaping) character etc.

**Comment by elfmz** [comment]
However ctrl+numpad digits works fine..

**Comment by unxed** [comment]
> Check this code: https://www.paste.org/120735

MATE terminal.

Ctrl+1..Ctrl+0 generates:

1\x00 \x1b\x1c\x1d\x1e\x1f\x7f90

No difference between right and left control. Numpad works ok.

Ctrl+0 was intercepted by terminal by default, had to disable it.

**Comment by unxed** [comment]
kde's konsole behaves just the same way. looks like it is standard de facto:
https://unix.stackexchange.com/questions/226327/what-does-ctrl4-and-ctrl-do-in-bash

btw, why not to convert whose to ctrl+digits in tty mode if we definitely know that ctrl is pressed?

**Comment by unxed** [comment]
`\x00` for `Shift+Ctrl+2`
and
`\x1e` for `Shift+Ctrl+6`
also

**Comment by unxed** [comment]
> However ctrl+numpad digits works fine..

ctrl+shift+numpad does not

**Comment by elfmz** [comment]
Its not [SPACE]\x1b but simple \x1b i.e. beginning of ESC sequence and cannot be detected separately (space that you see in report of termtest is just for better visibility of ESC sequences.
Also Ctrl+\ sends 0x1c just like Ctrl+3 (and far2l decodes it into Ctrl+\)
Some characters not decoded alternatively at least now, but still full set seems is unreachable due to conflicts above..

**Comment by unxed** [comment]
There is X key logging tool, [xkbcat](https://github.com/anko/xkbcat/blob/master/xkbcat.c). Can we use the same method as it does for getting entire keypress info from X in ttyx mode?

Okay, to implement such behavior, we need to know if focus is on our window. So we can assume we have focus if we receive any character in terminal, and we can assume focus is lost if we get any non-modifier keypress from X and in terminal we see nothing.

**Comment by elfmz** [comment]
oh i already see funny and weird bugs :) but it worth to try
regarding bugs, i tried it and see sometimes duplicated characters, what worse - in case of switching to russian that dups having russiand + corresponding latin presses:
```
Shift_L
f
ff
ff
ff
ff
fAlt_L
Shift_L
g
пg
пg
пg
пg
пg
пg
пControl_L
Control_L
```

**Comment by unxed** [comment]
try with -up switch

it will show something like
```
+g
g-g
```
for english layout and
```
+g
п-g
```
for russian

so it looks like for any alphanumeric key it shows:
key down event
translated to selected layout char
key up event

**Comment by unxed** [comment]
Another advantage of this approach is the ability to respond to keystrokes that DE or terminal are trying to intercept.

We can even use that approach in wx backend to process Alt+russian_letter correctly on wx < 3.1 (as wx 3.0 can not handle it)

**Comment by unxed** [comment]
duplicated / translated to current layout characters are tty echo
run `stty -echo && ./xkbcat -up` to see correct output
(so it looks like we still need to process terminal input to get correct alphanumeric chars in current keyboard layout)

**Comment by unxed** [comment]
Xspy is the same tool as xkbcat, but with support for keyboard layout translations:
https://gitlab.com/kalilinux/packages/xspy/-/blob/kali/master/Xspy.c

But it always use key mapings for DEFAULT keyboard layout, not CURRENT one.

Maybe we still need to process terminal input to get correct alphanumeric chars in current keyboard layout. Or maybe there is some way to determine keyboard layout currently selected (system wide or for terminal window only).

**Comment by unxed** [comment]
Updated xkbcat with support for getting chars in current keyboard layout:

https://gist.github.com/unxed/f331e27191b7de12ffaa27d01a862f6d

**Comment by unxed** [comment]
Updated xkbcat with support for getting chars in current keyboard layout:

https://gist.github.com/unxed/f331e27191b7de12ffaa27d01a862f6d

Sample output for "qw" in Russian and English layouts:

```
Cyrillic_shorti
й
1738
24

Cyrillic_tse
ц
1731
25

Shift_L
ц
65505
50

q
q
113
24

w
w
119
25
```

Third line is X11 KeySym. Forth is X11 event keycode.

Needs `libxkbcommon-dev` also.

**Comment by unxed** [comment]
To get Virtual Key Codes from X11 keycodes we need some kind of translation table.
X11 keycodes are listed here (X11 server section):
https://gist.github.com/rickyzhang82/8581a762c9f9fc6ddb8390872552c250

**Comment by elfmz** [comment]
may just use this only for ctr/alt+ events, as language doesn't matter there

**Comment by unxed** [comment]
> may just use this only for ctr/alt+ events, as language doesn't matter there

language definitely matters for Alt+letter quick search

fortunately, updated xkbcat above works with multiple keyboard layouts just fine :)

**Comment by unxed** [comment]
actually original xkbcat algorithm is worse for us as it prints chars not from English keyboard layout, but from default keyboard layout

**Comment by unxed** [comment]
> To get Virtual Key Codes from X11 keycodes we need some kind of translation table. X11 keycodes are listed here (X11 server section): https://gist.github.com/rickyzhang82/8581a762c9f9fc6ddb8390872552c250

Actually, this table is partially incorrect according to my testing. So had to make my own one for a new xkbcat update. Now it shows Windows Virtual Key Codes also. All keys on my full standalone keyboard work ok with it.
https://gist.github.com/unxed/4ae473e78bfc30b7bf493a23eaeca077

**Comment by unxed** [comment]
It turns out you can just ask the terminal emulator to send special escape sequences on getting/losing focus:
https://unix.stackexchange.com/a/480138/422943

Useful for "X keyboard hook" input method

**Comment by unxed** [comment]
xkbcat was [updated](https://github.com/anko/xkbcat/issues/5) to support multiple keyboard layouts

**Comment by unxed** [comment]
another usability feature that needs «x hook input mode» is keypad selection keys: `+`, `-` and `*`

they do not work as expected in tty mode w/o extensions currently

**Comment by elfmz** [comment]
there is dedicated bug on that https://github.com/elfmz/far2l/issues/1140

**Comment by unxed** [comment]
patch to fix ctrl+digits
https://github.com/elfmz/far2l/issues/1140#issuecomment-997467935

**Comment by elfmz** [comment]
Well, as you can see in commit i decided to continue with minimalistic hooking approach where input events from XI hook 'extends' usual TTY input events, i.e. no need to track focus changes etc.
I believe it has minimum risk of compatibility problems, also introduces no extra delays. Regarding compatibility problems - some terminal applications has own macro features - like keypresses that maps to other keypresses or even to sequences. Using XI-only based input will bypass such features.
For now modifiers still are not 'live' but it still can be implemented, but will require focus events and not sure how to make it in current model.

**Comment by unxed** [comment]
So cool!

...building...

Well, if we use focus events only for modifiers suggestions, its probably not a big deal if network delays will cause bugs with it sometimes. At least no possible data loss.

As for bypassing macros, well, we can make this feature optional so user can decide himself, if he needs it or not.

**Comment by unxed** [comment]
Keypad selection keys and ESC works perfectly even without --ee, cool!
Still,
Ctrl+digits work only with keypad keys
Ctrl+Shift+digits not working at all

**Comment by unxed** [comment]
As for Ctrl+digits, those can be sent pass-through if some app is running inside internal terminal, and interpreted as far hotkeys in all other cases. That is the use case for sending ESC^ or ESC_ to far2l is totally unknown.

**Comment by elfmz** [comment]
Fix for Ctrl+Digits is quite obvious - just need to add more remap corrections, will do later. This will work for all digits except 0 because for some reason Ctrl+0 doesn't cause any events in my Gnome terminal, however in xterm it works perfectly and doesn't need any corrections. Also gonna fix some new issues with latching modifiers.

**Comment by unxed** [comment]
I tried it on a slow connection - the one on which I tested my patch. Alt + Tab breaks the input (

**Comment by unxed** [comment]
hm, `Ctrl+0` works just as plain `0` in my MATE Terminal which AFAIK is fork of GNOME Terminal — according to `showkey -a`

**Comment by elfmz** [comment]
Ha, its cuz it just used by gnome terminal shortcuts, i.e. configurable:
![image](https://user-images.githubusercontent.com/20957349/147484187-7cf3fb26-a660-4b6a-9317-5399004388e5.png)

**Comment by elfmz** [comment]
now should be better

**Comment by unxed** [comment]
first keypad `+` press works as plus sign, not selection key, on connection with long roundtrip time (this behavior was the same with previous version also)

but Alt-Tab does not break input any more, hooray!

**Comment by elfmz** [comment]
Try to increase 10 to 20..30 here: https://github.com/elfmz/far2l/blob/master/WinPort/src/Backend/TTY/TTYX/TTYX.cpp#L306

**Comment by unxed** [comment]
increasing 10 to 20 or 30 didn't solves 'first keypress issue' for me. even 100 not helps.

ctrl+digits are working, great!

**Comment by elfmz** [comment]
If increasing delay dont help then likely TTY event comes after corresponding Xi keyrelease arrived. Try keeping '+' key pressed little bit longer :)

**Comment by elfmz** [comment]
Likely need to mention in help that Xi doesn't work fine on slow connections..

**Comment by unxed** [comment]
> Likely need to mention in help that Xi doesn't work fine on slow connections..

Yep. Preferred way to use remote far2l in favourite terminal emulator in case of slow/unstable connection is to run local far2l with Xi in favourite terminal and do ssh user@example.com inside to use far2l extensions for remote input.

**Comment by elfmz** [comment]
Also there is issue with security. Not related to far2l, but for X forwarding. It appears that running ssh -X on arbitrary host gives that host essential ability to keylog your system:) Sometimes it doesn't matter, sometimes it DOES:)

**Comment by unxed** [comment]
Btw, let's keep this ticket open for some time in case some issues arise with new implementation — not to open new ones

**Comment by elfmz** [comment]
Check plz current master on your slow connection, i've made some improvements in handling laggy events..

**Comment by unxed** [comment]
building

**Comment by unxed** [comment]
'first keypress' issue (keypad `-+*` printing chars to command line on first press) now happens relatively rare (1 of 10 tries or even less often), great!

ctrl+digits still need multiple tries to work sometimes, but looks like this happens less often also

**Comment by unxed** [comment]
> TTY|F backend [...] It provides UX identical to GUI backend

It is more correct, nevertheless, to say that the provided UX is equivalent to the UX of the backend of the host version of far2l. If host far2l uses plain TTY backend, user should not expect GUI far2l experience.

**Comment by unxed** [comment]
played with XI_KEYDOWN_MAXWAIT_MSEC a little bit. value of 1000 gives around 98% probability of correct keystroke detection on slow, unstable channel

**Comment by elfmz** [comment]
may be it even can be autodetectable, plz try following experiment: add in the beginning and end of TTYX's constructor: 
```
TTYX(bool allow_xi)
{
auto start_time = std::chrono::steady_clock::now();
.... code
auto end_time = std::chrono::steady_clock::now();
fprintf(stderr, "!!! %lu\n", (long)std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count());
}
```
and then run with export FAR2L_STD=far2l.log and see how much time it usually takes on your slow connection
PS C++ committee gone unleashed mad with std::chrono

**Comment by unxed** [comment]
It's between 308 and 414 now. But it looks like both wired and mobile internet connections are not overloaded now, as everybody now drink/sleep/celebrate instead of generating traffic. So looks like I should repeat the test few hours (days?) later.

Meanwhile, I want to wish you and all the colleagues in the development of far2L a Happy New Year. Far is significantly important tool in my work, and I am incredibly glad that now I can use it without Wine.

**Comment by elfmz** [comment]
also, what RTT reports usual ping utility at same time to same host?

**Comment by unxed** [comment]
tested one more time, both wired and mobile connection (looks like i accidentally did not log actual mobile connection values on the previous try). now log "!!!" values are from 309 to 1416. ping from 2.77 ms to 333 ms (most of times around 30-50 ms).

**Comment by unxed** [comment]
btw, setting XI_KEYDOWN_MAXWAIT_MSEC to 1500 made input working in 100% cases! sometimes with delays but no buggy input at all!

**Comment by elfmz** [comment]
Added some autodetection logic..

**Comment by unxed** [comment]
no input bugs now, great!

**Comment by unxed** [comment]
Another option to grab full keypress information even without X is to use `/dev/input/by-path/platform-i8042-serio-0-event-kbd`. Here is demo from stackoverflow showing this method. Needs sudo, but why not to use it is we already have sudo permissions, for example?

```
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <linux/input.h>
#include <string.h>
#include <stdio.h>

static const char *const evval[3] = {
    "RELEASED",
    "PRESSED ",
    "REPEATED"
};

int main(void)
{
    const char *dev = "/dev/input/by-path/platform-i8042-serio-0-event-kbd";
    struct input_event ev;
    ssize_t n;
    int fd;

    fd = open(dev, O_RDONLY);
    if (fd == -1) {
        fprintf(stderr, "Cannot open %s: %s.\n", dev, strerror(errno));
        return EXIT_FAILURE;
    }
    while (1) {
        n = read(fd, &ev, sizeof ev);
        if (n == (ssize_t)-1) {
            if (errno == EINTR)
                continue;
            else
                break;
        } else
        if (n != sizeof ev) {
            errno = EIO;
            break;
        }
        if (ev.type == EV_KEY && ev.value >= 0 && ev.value <= 2)
            printf("%s 0x%04x (%d)\n", evval[ev.value], (int)ev.code, (int)ev.code);

    }
    fflush(stdout);
    fprintf(stderr, "%s.\n", strerror(errno));
    return EXIT_FAILURE;
}
```

**Comment by unxed** [comment]
grabbing keyboard input under wayland:
https://github.com/Aishou/wayland-keylogger

**Comment by elfmz** [comment]
I have ubuntu 20.04 under hands and when booted in wayland mode (and as prove - $XDG_SESSION_TYPE == "wayland") it still sets up $DISPLAY env and provides X11 abilities through it. So it seems it least for now can just dont care about wayland in this aspect :)

**Comment by unxed** [comment]
Asked on stackoverflow if anyone know some trick to get detailed keypress info in Linux Terminal:
https://stackoverflow.com/questions/70673452/is-there-any-way-to-distinguish-keypresses-of-on-numeric-keypad-and-on

**Comment by unxed** [comment]
> It turns out you can just ask the terminal emulator to send special escape sequences on getting/losing focus:

Found better way of determining do terminal window have focus of not even if terminal does not sends onfocus/onblur escape sequences.

1) get X11 window id
2) ask X11 about focus state.

Trick to get X11 window id:
```
#!/bin/bash
uuid=$(uuidgen)
echo -ne "\033]0;${uuid}\007"
pid=$(xdotool search --name ${uuid})
echo $pid
read
```

**Comment by unxed** [comment]
> It turns out you can just ask the terminal emulator to send special escape sequences on getting/losing focus: https://unix.stackexchange.com/a/480138/422943

Focus tracking implemented in #2567
